### DISCUSSION TITLE
Why is the Python relational API faster than the DB API?

### DISCUSSION THREAD

**Discussion by jcrist** [duckdb/duckdb#8582] [General]
A duckdb performance question came up in the ibis issue tracker (https://github.com/ibis-project/ibis/issues/6874). After a bit of debugging, I've reduced this down to the following example:

<details>
<summary>Example code</summary>

```python
from contextlib import contextmanager
import timeit

import pandas as pd
import numpy as np
import duckdb


@contextmanager
def timer(label):
    start = timeit.default_timer()
    yield
    end = timeit.default_timer()
    elapsed = end - start

    print(f"{label}: {elapsed:.2f} seconds.")


N_PRICES = 5_000_000
N_TRANSACTIONS = 2_000_000


prices = pd.DataFrame(
    {
        "time": pd.date_range(start="09:00", periods=N_PRICES, freq="T"),
        "price": np.linspace(0, N_PRICES, N_PRICES)
    }
)
transactions = pd.DataFrame(
    {
        "time": np.random.choice(
            pd.date_range(
                start="09:00", periods=N_TRANSACTIONS * 2, freq="30S"
            ),
            size=N_TRANSACTIONS,
            replace=False,
        ),
        "shares": np.random.randint(5, 20, size=N_TRANSACTIONS)
    }
)

con = duckdb.connect()
con.execute("CREATE TABLE transactions AS SELECT * FROM transactions")
con.execute("CREATE TABLE prices AS SELECT * FROM prices")

sql = """
SELECT
    t.time, t.shares, p.price
FROM
    transactions t ASOF LEFT JOIN prices p
ON
    t.time >= p.time
"""

with timer("Using sql and to_arrow_table"):
    con.sql(sql).to_arrow_table()


with timer("Using execute and fetch_arrow_table"):
    con.execute(sql).fetch_arrow_table()
```

</details>

Results:

```
$ python bench.py
Using sql and to_arrow_table: 3.26 seconds.
Using execute and fetch_arrow_table: 10.61 seconds.
```

In short - using `con.sql(...).to_arrow_table()` appears to be measurably faster than `con.execute(...).fetch_arrow_table()`. Given the ratio I'm assuming there's parallelism going on somewhere in the former that is missing in the latter.

I looked through the docs and issue tracker and couldn't find any reference explaining this performance difference.

- Is is known that these methods differ? Or is this a performance bug?
- Is there anything that could be done to accelerate the `con.execute(...).fetch_arrow_table()` code path to close this performance gap?

**Comment by Tishj** [comment]
`sql` creates a logical plan, this is not directly executed
Only once the result is fetched, do we execute it, and that also means we can execute this in a more performant matter because we know what the result destination is.

`execute` does what it says, directly execute the query, the discrepancy you're noticing is likely related to the result materialization from `execute` followed by the conversion cost of this materialized result to an arrow table

**Reply by Tishj** [reply]
Also `fetch_arrow_table`, `to_arrow_table` and `arrow` are the same method, just aliases

**Reply by jcrist** [reply]
> Only once the result is fetched, do we execute it, and that also means we can execute this in a more performant matter because we know what the result destination is.

Thanks, that makes sense!

**Reply by jcrist** [reply]
This came up again - since duckdb 0.9.0 I can no longer detect a meaningful performance difference between using `con.sql(...).to_arrow_table()` and `con.execute(...).fetch_arrow_table()` (at least given the above benchmark). Are there still situations where we'd expect one to be faster? Or should these now be interchangeable performance-wise?

**Reply by Tishj** [reply]
They're not expected to be interchangable and will likely diverge in the future again
